src/content/en/2021/resource-hints.md

```diff
@@ -1,20 +1,609 @@
+
+#### `preload`
+
+Most modern browsers also <a hreflang="en" href="https://caniuse.com/link-rel-preload">support</a> the `preload` hint—and to a <a hreflang="en" href="https://caniuse.com/link-rel-modulepreload">lesser degree</a>, the <a hreflang="en" href="https://html.spec.whatwg.org/multipage/links.html#link-type-modulepreload">`modulepreload`</a> hint. The `preload` instruction initiates an early fetch for a resource which is required in the loading of a page and is most commonly used for late-discovered resources, such as font files or images referenced in stylesheets. Preloading a resource may be used to elevate its priority, allowing the developer to prioritize the loading of the <a hreflang="en" href="https://web.dev/lcp">Largest Contentful Paint</a> image for, even if this would otherwise be discovered while parsing the HTML.
+
+`modulepreload` is a specialized alternative to `preload` and behaves similarly, however its usage is limited to <a hreflang="en" href="https://html.spec.whatwg.org/multipage/webappapis.html#module-script">module scripts</a>.
+
+## Adoption and trends
+
+{{ figure_markup(
+  image="resource-hints-adoption.png",
+  caption='Adoption of the link rel attribute.',
+  description='Bar chart showing the percent of pages using the link rel attribute values. `dns-prefetch` has 36.4% adoption on mobile and 35.7% on desktop. `preload` has 22.1% adoption on mobile and 22.0% on desktop. `preconnect` has 12.7% adoption on mobile and 12.9% on desktop. `prefetch` has 2.1% adoption on mobile and 2.4% on desktop. `prerender` has 0.1% adoption on mobile and 0.1% on desktop. `modulepreload` has 0.1% adoption on mobile and 0.1% on desktop.',
+  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1740212588&format=interactive",
+  sheets_gid="2077755325",
+  sql_file="hints_adoption.sql"
+  )
+}}
+
+The most widely used resource hint is `dns-prefetch` (36.4% on mobile); which is unsurprising, considering it was introduced in <a hreflang="en" href="https://caniuse.com/link-rel-dns-prefetch">2009</a>. With the widespread use of HTTPS, in many cases it should be replaced with `preconnect` (12.7% on mobile). Considering that the `preload` hint is comparatively new, first appearing in Chrome in <a hreflang="en" href="https://groups.google.com/a/chromium.org/g/blink-dev/c/_nu6HlbNQfo/m/XzaLNb1bBgAJ?pli=1">2016</a>, it is the second most widely adopted resource hint (22.1% on mobile) and is seeing constant growth year-on-year—a testament to the importance and flexibility of this directive.
```

Comment:
If `preconnect` is preferable to `dns-prefetch` this sounds like an opportunity to improve the ecosystem by identifying who the most influential entities are for `dns-prefetch` adoption. Too late to get this analysis into the chapter so I spun up this discussion on the forum: https://discuss.httparchive.org/t/who-is-driving-dns-prefetch-adoption/2263

Reply:
Update: some responses to the thread have suggested that maybe there are still good reasons to keep `dns-prefetch` since `preconnect` is more expensive.

Is the "in many cases" of this paragraph doing too much heavy lifting? Stronger language might be needed to clarify when one hint might be more appropriate than the other.

Reply:
Originally I had written,
> in many cases it _could_ be replaced
And then changed it to _should_.

Personally I think both have their uses and it isn't like-for-like, but I do believe that in many cases (not all) `dns-prefetch` should be replaced with `preconnect`.

When introducing the `preconnect` hint, I say:
> You are able to use `preconnect` in place of `dns-prefetch` as it gives a greater performance boost; but you must use it sparingly as certificates are usually upwards of 3 KB, which would be competing with bandwidth for other resources. You also want to avoid wasting CPU time opening connections which aren't required for critical resource. Keep in mind that if a connection isn't used within a short period of time (e.g., 10 seconds on Chrome), it would automatically be closed by the browser, wasting any `preconnect` effort.

I was hoping it would help clarify when it should be replaced and when it should not.

Let me know what you think. I agree with the comments on the thread and am trying to convey the same thing.

Reply:
Given your analysis on that thread it seems both Wordpress and Google Fonts use cases would benefit from this. Maybe you could mention those examples? And then add a caveat about its expense (“as discussed in the pre connection section, there is an added risk/expense…”).

Could also link to that thread?

Reply:
Updated to:

> With the widespread use of HTTPS, in many cases you should replace it with `preconnect` (12.7% on mobile), if you are certain that you will be connecting to that domain.

Didn't want to link to the thread since the content is already in the chapter.